### PR TITLE
Smart contracts integration

### PR DESCRIPTION
**Pull request by mlopezdetrinidad**

### CODE REVIEW

File: pages/agreements.tsx [hunk 1]
```diff
@@ -1,35 +1,118 @@
...
+                  types.push(properties[objKey].type);
+                }
+                if (objKey === COUNTER_PARTY_NAME_FIELD) {
+                  counterPartyNameIndex = propsIndex;
+                }
+                propsIndex += 1;
+              });
+            });
+            const values = ethers.utils.defaultAbiCoder.decode(types, metadata);
+            console.log('meta ==>', values);
```
Comment (zubietaroberto) [inline pages/agreements.tsx:92]: Remove Console

File: pages/new-agreement.tsx [hunk 2]
```diff
@@ -179,49 +194,76 @@ const NewAgreement: NextPage<NewAgreementProps> = ({ templateTypeCode }) => {
...
-        agreementFormTemplateId: templateTypeCode,
-        escrowed: null,
-        validUntil: '12/21/2023',
-        toSigner: null,
-        fromSigner: null,
-        fileString: renderToString(agreementTemplate()),
-      },
-    };
+      const fee = await contract.fee();
+      const escrowAddress = await contract.escrow();
```
Comment (zubietaroberto) [inline pages/new-agreement.tsx:238]:
```suggestion
      const [fee, escrowAddress] = await Promise.all([contract.fee(), contract.escrow()]);
```

File: pages/profile.tsx [hunk 3]
```diff
@@ -3,12 +3,19 @@ import { useDispatch, useSelector } from 'react-redux';
...
+import { DID } from 'dids';
+import { arrayify, mnemonicToSeed } from 'ethers/lib/utils';
 import ProfileStateModel from '../models/profileStateModel';
 import FormProfile from '../components/profile/FormProfile';
 import PassphraseModal from '../components/profile/PassphraseModal';
 import ProfileModel from '../models/profileModel';
 import doSetProfile from '../redux/actions/profile';
 
+const elliptic_1 = require('elliptic');
+const did_jwt_1 = require('did-jwt');
```
Comment (zubietaroberto) [inline pages/profile.tsx:17]: Please use `import`

File: pages/profile.tsx [hunk 4]
```diff
@@ -31,34 +38,51 @@ const Profile: FC = () => {
...
+            const kpInstance = kp.keyFromSecret(keystore.keypairs.ED25519);
+            const walletAddress = did_jwt_1.toEthereumAddress(kpInstance.getPublic('hex'));
+
+            /* const walletDid = await xdvWallet.createES256K({
               rpcUrl: process.env.NEXT_PUBLIC_RPC_URL,
               walletId: profileData.walletId,
               registry: '',
               accountName: profileData.accountName,
-            });
+            }); */
```
Comment (zubietaroberto) [inline pages/profile.tsx:79]: Remove comment

File: pages/profile.tsx [hunk 5]
```diff
@@ -90,30 +114,40 @@ const Profile: FC = () => {
...
+      const accountName = values.name.replace(' ', '').toLowerCase();
       const xdvWallet = new Wallet({ isWeb: true });
+
       await xdvWallet.open(accountName, values.passphrase);
+
       await xdvWallet.enrollAccount({
         passphrase: values.passphrase,
         accountName,
       });
+      console.log(xdvWallet);
```
Comment (zubietaroberto) [inline pages/profile.tsx:126]: Remove Log

File: pages/profile.tsx [hunk 6]
```diff
@@ -90,30 +114,40 @@ const Profile: FC = () => {
...
+      const walletAddress = did_jwt_1.toEthereumAddress(kpInstance.getPublic('hex'));
+
+      /* const walletDid = await xdvWallet.createES256K({
         passphrase: values.passphrase,
         rpcUrl: process.env.NEXT_PUBLIC_RPC_URL,
         walletId,
         registry: '',
         accountName: values.name,
-      });
+      }); */
```
Comment (zubietaroberto) [inline pages/profile.tsx:142]: Remove Comment

File: redux/actions/agreement.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
 import AgreementModel from '@/models/agreementModel';
-import agreementsData from '../../data/agreements';
+// import agreementsData from '../../data/agreements';
```
Comment (zubietaroberto) [inline redux/actions/agreement.ts:2]: Remove Comment

**Review by zubietaroberto — CHANGES_REQUESTED** [review]

**Review by rubengue — DISMISSED** [review]
great work

File: package.json [hunk 8]
```diff
@@ -12,17 +12,25 @@
...
     "@hookform/error-message": "0.0.5",
     "@rjsf/core": "^2.4.2",
     "babel-plugin-transform-define": "^2.0.0",
     "bootstrap": "^4.6.0",
     "classnames": "^2.2.6",
     "date-fns": "^2.19.0",
+    "dids": "^2.1.0",
+    "eslint": "^6.8.0",
+    "ethers": "^5.3.1",
+    "ganache": "^6.4.4",
```
Comment (molekilla) [inline package.json:26]: dev deps

File: pages/agreements.tsx [hunk 9]
```diff
@@ -1,35 +1,118 @@
...
+        const filterRec = contract.filters.SignatureRequested(null, null, account);
+        const eventsRec = await contract.queryFilter(filterRec);
+        const myEvents = [...events, ...eventsRec];
+        const newAgreements = [];
+        const documents = myEvents.map((myEvent) => contract.anchors(myEvent.args.documentId));
+        const forkedDocuments = forkJoin(documents).pipe(debounce((x) => x as any)).toPromise();
+        const myDocuments = (await forkedDocuments) || [];
+        const ipfsItems = myDocuments.map((myDocument: any) => ipfs.getObject(myDocument.fileHash));
+        const forkedIpfsItems = forkJoin(ipfsItems).pipe(debounce((x) => x as any)).toPromise();
+        const myIpfsItems = (await forkedIpfsItems) || [];
```
Comment (molekilla) [inline pages/agreements.tsx:66]: we'll migrate to a gql lib later

File: pages/new-agreement.tsx [hunk 10]
```diff
@@ -179,49 +194,76 @@ const NewAgreement: NextPage<NewAgreementProps> = ({ templateTypeCode }) => {
...
-    }
+  const onSubmitForm = async () => {
+    try {
+      const ipfsManager = new IPLDManager(did);
+      await ipfsManager.start(process.env.NEXT_PUBLIC_IPFS_URL);
+      const fil = Buffer.from(renderToString(agreementTemplate()));
+      const cid = await ipfsManager.addSignedObject(fil,
+        {
+          name: agreementTitle,
+          contentType: '',
```
Comment (molekilla) [inline pages/new-agreement.tsx:205]: text/html

**Review by molekilla — CHANGES_REQUESTED** [review]

File: pages/agreements.tsx [hunk 11]
```diff
@@ -77,10 +77,13 @@ const Agreements: React.FC = () => {
             jsonSchemas.forEach((jsonSchema) => {
               const { properties } = jsonSchema;
               Object.keys(properties).forEach((objKey) => {
-                if (properties[objKey].custom === 'address') {
-                  types.push('address');
+                // if (properties[objKey].custom === 'address') {
+                //   types.push('address');
+                // }
```
Comment (zubietaroberto) [inline pages/agreements.tsx:82]: Remove Comments

**Review by zubietaroberto — CHANGES_REQUESTED** [review]
One final change

**Review by zubietaroberto — APPROVED** [review]
Approved

**Review by molekilla — APPROVED** [review]
lg

**Pull request merged**